### PR TITLE
ENH: Index optional pivot

### LINKED ISSUE #3962: Make DataFrame.pivot index argument truly optional
Creating a pivot table from a DataFrame is easy.

```python
In [17]: df
Out[17]: 
                 date variable     value
0 2000-01-03 00:00:00        A  0.528219
1 2000-01-04 00:00:00        A -0.135071
2 2000-01-05 00:00:00        A -0.343018
3 2000-01-03 00:00:00        B -0.097701
4 2000-01-04 00:00:00        B -1.383248

In [18]: df.pivot('date', 'variable', 'value')
Out[18]: 
variable           A         B
date                          
2000-01-03  0.528219 -0.097701
2000-01-04 -0.135071 -1.383248
2000-01-05 -0.343018       NaN
```

However if the index has been set, and there is no column to be set as index, pivot fails. From api point of view, the ```index``` argument is optional, but in fact it is not.

```python
In [19]: df.set_index('date', inplace=True)

In [20]: df
Out[20]: 
           variable     value
date                         
2000-01-03        A  0.528219
2000-01-04        A -0.135071
2000-01-05        A -0.343018
2000-01-03        B -0.097701
2000-01-04        B -1.383248

In [21]: df.pivot(columns='variable', values='value')
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
...
KeyError: u'no item named None'
````

Of course one can reset the index before using pivot.

```python
In [22]: df.reset_index().pivot('date', 'variable', 'value')
Out[22]: 
variable           A         B
date                          
2000-01-03  0.528219 -0.097701
2000-01-04 -0.135071 -1.383248
2000-01-05 -0.343018       NaN
```

### PR DESCRIPTION
**Pull request by jseabold**
Closes #3962.

### CODE REVIEW

File: pandas/core/reshape.py [hunk 1]
```diff
@@ -319,12 +319,20 @@ def pivot(self, index=None, columns=None, values=None):
...
-        indexed = self.set_index([index, columns])
+        cols = [columns] if index is None else [index, columns]
+        append = True if index is None else False
+        indexed = self.set_index(cols, append=append)
         return indexed.unstack(columns)
     else:
-        indexed = Series(self[values].values,
-                         index=MultiIndex.from_arrays([self[index],
-                                                       self[columns]]))
+        if index is None:
```
Comment (jreback) [inline pandas/core/reshape.py]: just do:

```
if index is None:
   index = self.index
else:
   index = self[index]

indexed = Series(..........)
```

File: pandas/core/reshape.py [hunk 2]
```diff
@@ -319,12 +319,20 @@ def pivot(self, index=None, columns=None, values=None):
     See DataFrame.pivot
     """
     if values is None:
-        indexed = self.set_index([index, columns])
+        cols = [columns] if index is None else [index, columns]
+        append = True if index is None else False
```
Comment (jreback) [inline pandas/core/reshape.py]: ``append = index is None``

File: pandas/tests/test_frame.py [hunk 3]
```diff
@@ -9478,6 +9478,46 @@ def test_pivot_integer_bug(self):
         repr(result)
         self.assert_numpy_array_equal(result.columns, ['A', 'B'])
 
+    def test_pivot_index_none(self):
+        data = {
```
Comment (jreback) [inline pandas/tests/test_frame.py]: add the issue number as a comment

**Comment by jseabold** [comment]
Addressed comments.

**Comment by jseabold** [comment]
Failures look unrelated?

**Comment by jreback** [comment]
lgtm. can you add a release note refing the original issue. squash, then ping when green.

**Comment by jseabold** [comment]
Done

**Comment by jreback** [comment]
- can you update the doc-strings of both ``pd.pivot`` and ``DataFrame.pivot`` (saying ``index`` is optional/can be ``None``).
- can you add a test using ``pd.pivot``?

**Comment by jseabold** [comment]
Are there any tests for `pd.pivot`? It's just a pass-through.

**Comment by jreback** [comment]
looks ``pd.pivot_table == DataFrame.pivot_table``. While ``pd.pivot == DataFrame.pivot``.

So ok on ``.pivot``. Maybe have a look at ``.pivot_table``?

**Comment by jseabold** [comment]
It appears to work.

```
In [8]: df
Out[8]:
     A    B      C  D
0  foo  one  small  1
1  foo  one  large  2
2  foo  one  large  2
3  foo  two  small  3
4  foo  two  small  3
5  bar  one  large  4
6  bar  one  small  5
7  bar  two  small  6
8  bar  two  large  7

In [9]: pd.pivot_table(df, values='D', index=None, columns=['C'], aggfunc=np.sum)
Out[9]:
C
large    15
small    18
Name: D, dtype: int64

In [10]: pd.pivot_table(df, values='D', index=['A', 'B'], columns=['C'], aggfunc=np.sum)
Out[10]:
C        large  small
A   B
bar one      4      5
    two      7      6
foo one      4      1
    two    NaN      6
```

**Comment by jreback** [comment]
ok gr8. test pretty pretty pls.

**Comment by jseabold** [comment]
It's tested. https://github.com/pydata/pandas/blob/master/pandas/tools/tests/test_pivot.py#L60

**Comment by jreback** [comment]
ic, I guess that already accepted ``None`` then, ok.

**Comment by jreback** [comment]
ok, just fix the doc-string and push again.

File: doc/source/whatsnew/v0.17.0.txt [hunk 4]
```diff
@@ -110,6 +110,8 @@ Other enhancements
 
 - ``pd.merge`` will now allow duplicate column names if they are not merged upon (:issue:`10639`).
 
+- ``pd.pivot`` will now allow passing index as None (:issue:`3962`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.17.0.txt]: use backticks on ``None``

**Comment by jseabold** [comment]
Done.

**Comment by jreback** [comment]
ok cool. ping when green.

**Comment by jreback** [comment]
thank you sir!

**Pull request merged**